### PR TITLE
[Inductor max autotune] Multithreaded Precompilation

### PR DESCRIPTION
**Pull request by kadeng**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115877
* #115814
* #115813
* #115655
* #115654
* #112861
* #114687
* #115270
* #115174
* #115072
* #114606
* #114319
* #114075
* #113932
* #113891
* #113959
* #113890
* #113570
* #113569
* __->__ #113558
* #113399
* #113366
* #113365

When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/select_algorithm.py [hunk 1]
```diff
@@ -774,7 +780,26 @@ def __call__(
...
+                return
+            with ThreadPoolExecutor(
+                max_workers=min(
+                    config.autotune_precompilation_workers, torch.get_num_threads()
+                )
+            ) as executor:
+                executor.map(
+                    lambda c: c.precompile(),
+                    [c for c in choices if hasattr(c, "precompile")],
+                    timeout=60 * 15,
```
Comment (aakhundov) [inline torch/_inductor/select_algorithm.py:794]: Maybe make this a constant?

File: torch/_inductor/select_algorithm.py [hunk 2]
```diff
@@ -774,7 +780,26 @@ def __call__(
...
+                )
+
         def autotune(choices):
+            try:
+                precompile(choices)
+            except TimeoutError:
+                log.warning("Precompilation phase took longer than timeout allowed. Continuing")
```
Comment (aakhundov) [inline torch/_inductor/select_algorithm.py:801]: What happens in precompilation times out? Is the subsequent `make_benchmark_fn()(choices)` expected to run fine?

File: torch/_inductor/select_algorithm.py (hunk 1 above)
Comment (kadeng) [reply torch/_inductor/select_algorithm.py]: Made it a parameter that's passed at construction..

File: torch/_inductor/select_algorithm.py (hunk 2 above)
Comment (kadeng) [reply torch/_inductor/select_algorithm.py]: Yes, nothing bad happens in that case.

**Comment by kadeng** [comment]
Moved to a (draft) feature branch, see https://github.com/pytorch/pytorch/pull/115919

**Pull request closed without merging**